snapshotter/snapshotter/layer.go

```diff
@@ -0,0 +1,292 @@
+		} else {
+			// Create header.
+			header, err = tar.FileInfoHeader(info, info.Name())
+		}
+
+		if err != nil {
+			return fmt.Errorf("failed to create FileInfoHeader. %w", err)
+		}
+
+		// Remove the prefix (dirPath) from each path.
+		relativePath := path[len(dirPath):]
+		header.Name = relativePath
+
+		if path == dirPath {
+			// For root folder, set the modification time to default time.
+			header.Name = "/"
+			header.ModTime = defaultDate
+		}
+
+		// Set the access and change times to the modification time.
+		// This is essential for reproducibility.
+		// See: https://reproducible-builds.org/docs/archives/
+		header.AccessTime = header.ModTime
+		header.ChangeTime = header.ModTime
+
+		// Write the header
+		if err := w.WriteHeader(header); err != nil {
+			return fmt.Errorf("failed to write header. %w", err)
+		}
+
+		// Skip writing contents if not a regular file.
+		if !info.Mode().IsRegular() {
+			return nil
+		}
+
+		// Write the contents of the file to the tar archive.
+		file, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s for copying. %w",
+				path, err)
+		}
+		defer file.Close()
+
+		_, err = io.Copy(w, file)
+		if err != nil {
+			return fmt.Errorf("failed to copy %s. %w", path, err)
+		}
+		return err
+	})
+	return err
+}
+
+// Compute md5 hash of given data
+func computeMD5Hash(data []byte) [16]byte {
+	var sum [16]byte
+	hash := md5.Sum(data)
+	// Retype to [16]byte
+	copy(sum[:], hash[:16])
+	return sum
+}
+
+// Compute the dmverity merkeltree and append to the disk image file.
+func computeAndWriteHashDevice(diskImagePath string) (string, error) {
+	w, err := os.OpenFile(diskImagePath, os.O_RDWR, 0644)
+	if err != nil {
+		return "", fmt.Errorf("failed to open %s. %w", diskImagePath, err)
+	}
+
+	// Find the size of the disk image.
+	ext4blocksize := 4096
+	ext4size, err := w.Seek(0, io.SeekEnd)
+	if err != nil {
+		return "", fmt.Errorf("failed to seek end. %w", err)
+	}
+
+	// Rewind the stream and then read it all into a []byte for
+	// dmverity processing
+	_, err = w.Seek(0, io.SeekStart)
+	if err != nil {
+		return "", fmt.Errorf("failed to seek start. %w", err)
+	}
+	data, err := ioutil.ReadAll(w)
+	if err != nil {
+		return "", fmt.Errorf("failed to read all. %w", err)
+	}
+
+	// Compute merkeltree of the disk image
+	mtree, err := dmverity.MerkleTree(data)
+	if err != nil {
+		return "", fmt.Errorf("failed to build merkle tree. %w", err)
+	}
+
+	// Write dmverity superblock and then the merkle tree after the end of the
+	// ext4 filesystem
+	_, err = w.Seek(0, io.SeekEnd)
+	if err != nil {
+		return "", fmt.Errorf("failed to seek end for writing dmverity data. %w", err)
+	}
+
+	// Create a new dmverity superblock object.
+	superblock := dmverity.NewDMVeritySuperblock(uint64(ext4size))
+
+	// By default, a random UUID is generated. Use the md5 of data so that the UUID is
+	// reproducible.
+	superblock.UUID = computeMD5Hash(data)
```

Comment:
I thought the superblock UUID was a SHA-256 rather than an MD5. Does this implementation use MD5?